starlette/routing.py

```diff
@@ -572,6 +574,91 @@ def __call__(self: _T, app: object) -> _T:
         return self
 
 
+class LifespanException(Exception):
+    def __init__(self, message: str):
+        self.message = message
+
+
+@asynccontextmanager
+async def _app_lifespan(scope: Scope, app: ASGIApp) -> typing.AsyncIterator[None]:
+    startup_sent = Event()
+    startup_done = Event()
+
+    shutdown_init = Event()
+    shutdown_sent = Event()
+    shutdown_done = Event()
+
+    lifespan_supported = False
+    exception = None
+
+    async def receive() -> Message:
+        nonlocal lifespan_supported
+
+        lifespan_supported = True
+
+        if not startup_sent.is_set():
+            startup_sent.set()
+            return {"type": "lifespan.startup"}
+
+        elif startup_done.is_set() and not shutdown_sent.is_set():
+            await shutdown_init.wait()
+            shutdown_sent.set()
+            return {"type": "lifespan.shutdown"}
+
+        else:
+            raise RuntimeError("unexpected receive")
+
+    async def send(message: Message) -> None:
+        nonlocal exception, lifespan_supported
+
+        lifespan_supported = True
+
+        if startup_sent.is_set() and not startup_done.is_set():
+            if message["type"] == "lifespan.startup.complete":
+                pass
+            elif message["type"] == "lifespan.startup.failed":
+                exception = message.get("message", "")
+            else:
+                raise ValueError(f"unexpected type: {message['type']}")
+            startup_done.set()
+
+        elif shutdown_sent.is_set() and not shutdown_done.is_set():
+            if message["type"] == "lifespan.shutdown.complete":
+                pass
+            elif message["type"] == "lifespan.shutdown.failed":
+                exception = message.get("message", "")
+            else:
+                raise ValueError(f"unexpected type: {message['type']}")
+            shutdown_done.set()
+
+        else:
+            raise RuntimeError("unexpected send")
+
+    # This wrapper is needed because TaskGroup.start_soon does not like that
+    # App returns Awaitable instead of Coroutine
+    async def coro_app(scope: Scope, receive: Receive, send: Send) -> None:
+        await app(scope, receive, send)
+
+    try:
+        async with create_task_group() as tg:
+            tg.start_soon(coro_app, {**scope, "app": app}, receive, send)
```

Comment:
After all of the issues caused by using tasks in `BaseHTTPMiddelware` I'd like to avoid using them if possible. And I think that in this case it _is_ possible: https://github.com/adriangb/asgi-routing/blob/main/asgi_routing/_lifespan_dispatcher.py#L11-L101. It's not really more LOC either. But maybe there's a bug in there that I didn't catch.